Kernel/Bus/USB/UHCI/UHCIController.cpp

```diff
@@ -430,6 +430,50 @@ ErrorOr<size_t> UHCIController::submit_control_transfer(Transfer& transfer)
     return transfer_size;
 }
 
+ErrorOr<size_t> UHCIController::submit_bulk_transfer(Transfer& transfer)
+{
+    Pipe& pipe = transfer.pipe();
+    dbgln_if(UHCI_DEBUG, "UHCI: Received bulk transfer for address {}. Root Hub is at address {}.", pipe.device_address(), m_root_hub->device_address());
+
+    // Create a new descriptor chain
+    TransferDescriptor* last_data_descriptor;
+    TransferDescriptor* data_descriptor_chain;
+    auto buffer_address = Ptr32<u8>(transfer.buffer_physical().as_ptr());
+    TRY(create_chain(pipe, transfer.pipe().direction() == Pipe::Direction::In ? PacketID::IN : PacketID::OUT, buffer_address, pipe.max_packet_size(), transfer.transfer_data_size(), &data_descriptor_chain, &last_data_descriptor));
+
+    last_data_descriptor->terminate();
+
+    if constexpr (UHCI_VERBOSE_DEBUG) {
+        if (data_descriptor_chain) {
+            dbgln("Data TD");
+            data_descriptor_chain->print();
+        }
+    }
+
+    QueueHead* transfer_queue = allocate_queue_head();
+    if (!transfer_queue) {
+        free_descriptor_chain(data_descriptor_chain);
+        return 0;
```

Comment:
Shouldn't this be an `Error` return value instead?

Reply:
Possibly, the UHCI bulk transfer code is mostly borrowed from the preexisting UHCI control transfer code with a few tweaks as made sense. Would returning Error::from_errno(ENOMEM) be the correct way here?

Reply:
That's my guess as well!

Reply:
On second thought I may defer this one to @Quaker762, since any change here should also probably be matched in the control transfer code as well, and returning 0 here (indicating no data transferred) still sort of semantically makes sense from the point of the calling Pipe code which may want to retry the transfer until either all the data is transferred or some other condition is met (this is a check that also needs to be added for the Pipe::bulk_transfer at some point, I've neglected it for now since I'm not sure of the best way to approach it)

Reply:
Yeah this one may need a touch up after I'm done with driver probing things. From memory (which is extremely hazy because it's been quite some time since I wrote this code lol), this originally just returned a `size_t` and _not_ an `ErrorOr`. Realistically this should probably return `ENOMEM` because it failed to allocated a transfer queue correctly.